### ISSUE TITLE
Get the derivative of a node with respect to another one

### ISSUE THREAD

**Issue by SJ001** [pytorch/pytorch#21441]
Hello! How can I take the derivative of a node in my NN with respect to another one? Let's say I have this simple network:
```
class model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.l1 = torch.nn.Linear(2, 1, bias=False)
        self.l2 = torch.nn.Linear(1, 2, bias=False)

    def forward(self, x):
        z = self.l1(x)
        y_pred = self.l2(z)
        return y_pred, z
```

I want my loss to look like this (it is not syntactically correct from a python perspective): `loss = (y-y_pred)**2+abs(d(y_pred)/d(z)-1)`. So, beside the normal MSE loss I want to enforce the gradient of the output layer with respect to an inner layer to be 1. I was thinking of doing something like this (again it is just a scheme, not a python code):

```
loss_1 =  (y-y_pred)**2
loss_1.backward(retain_graph=True)
loss_2 = abs(d(y_pred)/d(z)-1)
loss_2.backward()
```
I think this would do what I want, but I am not sure how to get d(y_pred)/d(z). It is calculated when I call loss_1.backward(retain_graph=True), but I am not sure how to access it. Can someone help me please? Thank you!

**Comment by SsnL** [comment]
torch.autograd.gead

**Comment by SJ001** [comment]
> torch.autograd.grad

Thank you for your reply! Sorry, I am quite new to pytorch. Could you please elaborate a bit?

**Comment by fmassa** [comment]
You should pass `retain_graph` to `torch.autograd.grad`, as explained in the error message.
```python
l1 = torch.autograd.grad(y_pred[0], z, retain_graph=True)
l2 = torch.autograd.grad(y_pred[1], z, retain_graph=True)
```

Also, we use the GitHub issues to track bugs and feature requests. Questions like this are better suited to the forums https://discuss.pytorch.org
